File: 001/ruby/robots.rb
#!/usr/bin/env ruby
# encoding: utf-8

if (ARGV.length != 1) then
  puts "Usage: robots.rb <filename>"
  exit(1)
end

filename = ARGV[0]

def process rows, pos
  upper_west = rows[0][0..pos].reverse.map.with_index{|e, i| e*(1-i%2) }.reduce(:+)
  upper_east = rows[0][pos..-1].map.with_index{|e, i| e*(1-i%2) }.reduce(:+)
  lower_west = rows[1][0..pos].reverse.map.with_index{|e, i| e*(i%2) }.reduce(:+)
  lower_east = rows[1][pos..-1].map.with_index{|e, i| e*(i%2) }.reduce(:+)
  west = upper_west + lower_west
  east = upper_east + lower_east
  puts(west <= east ? "GO WEST" : "GO EAST")
end

rows = []
pos = nil
File.open(filename).each do |line|
  line.chomp!
  if line =~ /(.*)X/ then
    pos = $1.length
  elsif line=~ /^$/ then
    process(rows, pos)
    rows = []
  else
    rows << line.chars.map.with_index { |e, x| (e == "|" ? 1 : 0) }
  end
end
process(rows, pos)

File: 001/clojure/robots.clj
#!/usr/bin/env clj

(use 'clojure.java.io)

(if (nil? *command-line-args*)
  (do
    (println "Usage: clj robots.clj FILENAME")
    (System/exit 1)))

(def filename (first *command-line-args*))

(defn laser-test [char test]
  (if (and test (= char \|)) 1 0))

(defn rem-cycle [pos] 
  (drop pos (cycle '(true false))))

(defn process [[upper middle lower & r]]
  (let [pos (.indexOf middle "X")
        upper-west (map laser-test (take (inc pos) upper) (rem-cycle pos))
        upper-east (map laser-test (drop pos upper) (rem-cycle 0))
        lower-west (map laser-test (take (inc pos) lower) (rem-cycle (inc pos)))
        lower-east (map laser-test (drop pos lower) (rem-cycle 1))
        west (+ (reduce + upper-west) (reduce + lower-west))
        east (+ (reduce + upper-east) (reduce + lower-east))]
    (println (if (<= west east) "GO WEST" "GO EAST"))
    (if (not (nil? r )) (recur (rest r)))))


(with-open [rdr (reader filename)]
  (process (lazy-seq (line-seq rdr))))
